adapters/Vungle/VungleAdapter/Bidding/GADMediationVungleNativeAd.m

```diff
@@ -0,0 +1,248 @@
+// limitations under the License.
+
+#import "GADMediationVungleNativeAd.h"
+#include <stdatomic.h>
+#import "GADMAdapterVungleUtils.h"
+#import "GADMAdapterVungleRouter.h"
+#import <VungleSDK/VungleNativeAd.h>
+
+@interface GADMediationVungleNativeAd () <GADMediationNativeAd, VungleNativeAdDelegate, GADMAdapterVungleDelegate>
+
+@end
+
+@implementation GADMediationVungleNativeAd {
+  /// Ad configuration for the ad to be loaded.
+  GADMediationNativeAdConfiguration *_adConfiguration;
+
+  /// The completion handler to call when the ad loading succeeds or fails.
+  GADMediationNativeLoadCompletionHandler _adLoadCompletionHandler;
+
+  /// The Vungle native ad.
+  VungleNativeAd *_nativeAd;
+
+  /// The ad event delegate to forward ad rendering events to the Google Mobile Ads SDK.
+  id<GADMediationNativeAdEventDelegate> _delegate;
+    
+  /// The Vungle container for the main ad
+  VungleMediaView *_mediaView;
+}
+
+@synthesize desiredPlacement;
+@synthesize isAdLoaded;
+
+- (nonnull instancetype)initNativeAdForAdConfiguration:(nonnull GADMediationNativeAdConfiguration *)adConfiguration
+                                     completionHandler:(nonnull GADMediationNativeLoadCompletionHandler)completionHandler {
+  self = [super init];
+  if (self) {
+    _adConfiguration = adConfiguration;
+
+    // Store the ad config and completion handler for later use.
+    __block atomic_flag adLoadHandlerCalled = ATOMIC_FLAG_INIT;
+    __block GADMediationNativeLoadCompletionHandler originalCompletionHandler = [completionHandler copy];
+
+    // Ensure the original completion handler is only called once, and is deallocated once called.
+    _adLoadCompletionHandler = ^id<GADMediationNativeAdEventDelegate>(_Nullable id<GADMediationNativeAd> ad, NSError *_Nullable error) {
+      if (atomic_flag_test_and_set(&adLoadHandlerCalled)) {
+        return nil;
+      }
+      id<GADMediationNativeAdEventDelegate> delegate = nil;
+      if (originalCompletionHandler) {
+        delegate = originalCompletionHandler(ad, error);
+      }
+      originalCompletionHandler = nil;
+      return delegate;
+    };
+  }
+  return self;
+}
+
+- (void)requestAd {
+  self.desiredPlacement = [GADMAdapterVungleUtils findPlacement:_adConfiguration.credentials.settings
+                                                  networkExtras:_adConfiguration.extras];
+  if (!self.desiredPlacement) {
+    NSError *error = GADMAdapterVungleErrorWithCodeAndDescription(GADMAdapterVungleErrorInvalidServerParameters, @"Placement ID not specified.");
+    _adLoadCompletionHandler(nil, error);
+    return;
+  }
+    
+  if (![[VungleSDK sharedSDK] isInitialized]) {
+    NSString *appID = [GADMAdapterVungleUtils findAppID:_adConfiguration.credentials.settings];
+    [[GADMAdapterVungleRouter sharedInstance] initWithAppId:appID delegate:self];
+    return;
+  }
+    
+  [self loadAd];
+}
+
+- (void)loadAd {
+  _nativeAd = [[VungleNativeAd alloc] initWithPlacementID:self.desiredPlacement];
```

Comment:
I would assume if we support native bidding, bid response will be passed here correct?

Reply:
Yes. It should be set in the object initialization.